core/src/main/scala/cats/collections/Range.scala

```diff
@@ -103,14 +104,18 @@ final case class Range[A](val start: A, val end: A) {
   def contains(x: A)(implicit A: Order[A]): Boolean = A.gteqv(x, start) && A.lteqv(x, end)
 
   /**
-   * Apply function f to each element in range [star, end]
+   * Apply function f to each element in range `[star, end]`. Does nothing if `end` precedes `start`.
    */
   def foreach(f: A => Unit)(implicit discrete: Discrete[A], order: Order[A]): Unit = {
     var i = start
-    while (order.lteqv(i, end)) {
+    // Visit values in range [start, end), i.e. excluding `end`.
+    while (order.lt(i, end)) {
       f(i)
       i = discrete.succ(i)
     }
+    // Visit the last (or the only) value, if any.
+    if (order.eqv(i, end))
+      f(i)
```

Comment:
The previous implementation was resulting to infinite loops in cases like
```
Range(123.toByte, Byte.MaxValue).foreach(_ => ())
```
I.e. due to integer overflows that could not be caught previously.